동기화 메모장 [STEP 1] aCafelaCoffee, namu, zoe

aCafela-coffee opened the pull request:
안녕하세요 웨더! **@SungPyo**

수정이 필요한 부분들 한번씩 짚어주시면 감사하겠습니다 🙂
리뷰 잘 부탁드립니다~~ 🙏

## 고민한 점

### 세로 화면에서 사이드바 사용하기

```swift
// line 24, SceneDelegate.swift

let splitViewController = UISplitViewController(style: .doubleColumn)
splitViewController.preferredDisplayMode = .oneBesideSecondary
splitViewController.preferredSplitBehavior = .tile
```

세로 화면에서 .oneBesideSecondary로 보여주기 위해 .tile을 사용했습니다.

### 셀 삭제, 추가 애니메이션
<img src="https://user-images.githubusercontent.com/70484506/153402759-e5fea7ef-7d96-4f50-9292-abbef415e622.gif" width="50%" height="50%">

```swift
// 셀 추가
memos.insert(newMemo, at: 0)
tableView.insertRows(at: [firstCellIndexPath], with: .fade)
```

```swift
// 셀 삭제
memos.remove(at: indexPath.row)
tableView.deleteRows(at: [indexPath], with: .fade)
```

### 키보드가 디테일 뷰의 컨텐츠를 가리지 않을 것

키보드가 올라왔을 때 디테일 뷰의 내용을 가리지 않도록 텍스트뷰의 `contentInset` 값을 키보드의 높이만큼 더해주었습니다.

```swift
@objc private func keyboardWillShow(_ notification: Notification) {
  guard
    let userInfo = notification.userInfo,
    let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else {
      return
    }
  textView.contentInset.bottom = keyboardFrame.height
  textView.verticalScrollIndicatorInsets.bottom = keyboardFrame.height
}

@objc private func keyboardWillHide(_ notification: Notification) {
  textView.contentInset.bottom = 0
  textView.verticalScrollIndicatorInsets.bottom = 0
}
```

<img src="https://user-images.githubusercontent.com/70484506/153402461-5f80d9ed-58db-4081-87b2-2d7d00e948b7.gif" width="50%" height="50%">

### Memo.subtitle

TableViewCell 두번째 줄의 레이블은 NSAttributedString를 이용했습니다.

### 앱이 구동될 때 / 메모 삭제시 / 메모 추가 시 첫번째 셀 선택

앱이 처음 구동 될 때, 메모를 삭제할 때, 메모를 추가할 때 첫번째 셀을 선택하도록 구현했습니다. 사용자 입장에서 앱이 좀 더 편리하다고 느끼도록 했습니다.

### 메모 목록 VC와 메모 상세  VC 간 데이터 전달 방식

메모 목록을 table로 가지고있는 `MemoListViewController` 와, 메모의 내용을 표시하는 `DetailViewController` 간에 양방향으로 delegation 관계를 구현하여 데이터를 전달하도록 했습니다.

`MemoListViewController`는 테이블뷰 셀이 선택되면 `MemoDisplayable` 프로토콜을 채택한 `DetailViewController` 에게 Memo 인스턴스를 전달하여 화면에 표시할 데이터를 전해줍니다.

`DetailViewController`는 사용자가 메모의 내용을 수정하면 `MemoStorable`을 채택한 `MemoListViewController` 에게 변경된 Memo 인스턴스를 전달하여 수정한 내용이 반영되도록 합니다.

## 질문

### 셀 삭제시, 셀 선택이 취소됨

```swift
func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
  let deleteAction = UIContextualAction(style: .destructive, title: "Delete") { _, _, completionHandler in
    self.memos.remove(at: indexPath.row)
    if self.memos.isEmpty {
      self.addMemo()
    } else {
      // 셀 삭제
      tableView.deleteRows(at: [indexPath], with: .fade)
      
      // 가장 위의 셀 select
      let firstRowIndexPath = IndexPath(row: 0, section: 0)
      tableView.selectRow(at: firstRowIndexPath, animated: false, scrollPosition: .none)
      
      self.loadDetail(at: firstRowIndexPath)
    }
    completionHandler(true)
  }
  deleteAction.image = UIImage(systemName: "trash")
  let configuration = UISwipeActionsConfiguration(actions: [deleteAction])
  return configuration
}
```

테이블뷰의 셀을 스와이프하여 삭제한 후, 첫 번째 row의 cell을 select하는 기능을 위와 같이 구현하였지만 select되지 않는 문제가 있습니다.  

selectRow 메서드의 부가적인 기능인 scrollPosition은 동작하는데 cell selection이 작동하지 않는 원인이 무엇인지 아직 알아내지 못했습니다.

## 미해결

### 컬렉션뷰 및 DiffableDataSource 사용 실패

처음에 리스트 모양의 컬렉션뷰와 DiffableDataSource를 사용해서 MemoListViewController를 만들었습니다.
이때 수정한 메모를 저장하고, CollectionViewListCell에 반영하는 방법을 찾지 못했습니다.
[commit 786d1e1c4966df3ad5ffbc635964bcb5191fef47](https://github.com/yagom-academy/ios-cloud-notes/pull/80/commits/786d1e1c4966df3ad5ffbc635964bcb5191fef47) UITableView로 변경했습니다.

aCafela-coffee committed:
chore: SwiftLint 적용

aCafela-coffee committed:
chore: Targeted Device를 iPad만 설정

aCafela-coffee committed:
chore: iOS Deployment Target 14.0으로 변경

aCafela-coffee committed:
feat: SplitView 화면 추가

na-young-kwon committed:
chore: SwiftLint 설정 변경

na-young-kwon committed:
feat: Detail View UI 구현

aCafela-coffee committed:
feat: RootView UI 추가

na-young-kwon committed:
feat: RootViewCell 의 Subtitle 구현

- 메모의 날짜와 본문의 일부를 표시

na-young-kwon committed:
refactor: Memo.swift 에서 UIFont 만 import 하도록 수정

aCafela-coffee committed:
feat: 메모 추가 기능 구현

na-young-kwon committed:
refactor: RootViewCell의 subTitle의 body 글자수를 100자로 제한

na-young-kwon committed:
feat: 셀 선택시 메모를 보여주는 기능 추가

na-young-kwon committed:
refactor: 사용하지 않는 MemoCell.swift 파일 삭제

aCafela-coffee committed:
style: 줄바꿈 규칙 반영, if문 가독성 개선

jsim27 committed:
fix: Xcode12에서 컴파일 되지 않는 문법 수정

- guard let 문법 수정
- 그 외 부분적인 스타일 수정

jsim27 committed:
feat: Assets로부터 json파일 로드기능 구현

aCafela-coffee committed:
refactor: 날짜 표기 지역화

na-young-kwon committed:
feat: 스와이프로 메모삭제 기능 구현

na-young-kwon committed:
refactor: RootViewController 이니셜라이저 추가

jsim27 committed:
feat: 수정 기능 구현

jsim27 committed:
refactor: 타입명 수정

- RootViewController -> MemoListViewController
- DetailViewControllerDelegate -> MemoListViewControllerDelegate

aCafela-coffee committed:
refactor: 메모 수정을 위한 DetailViewControllerDelegate 사용

na-young-kwon committed:
refactor: MemoListViewController가 DiffableDataSource를 사용하도록 로직수정

jsim27 committed:
refactor: 스와이프로 메모 삭제기능 snapshot을 통해 적용되도록 수정

aCafela-coffee committed:
refactor: DiffableDataSource를 통한 메모 추가, 메모 열기 기능 구현

na-young-kwon committed:
style: Mark 주석 앞 공백 추가

na-young-kwon committed:
refactor: memoComponents 프로퍼티 분리

na-young-kwon committed:
refactor: 처음 메모를 저장하는 프로퍼티 삭제

jsim27 committed:
feat: 앱 구동 및 셀 추가시 셀 선택 기능 구현

aCafela-coffee committed:
refactor: MemoListView를 UITableView로 변경

na-young-kwon committed:
feat: 스와이프로 셀 삭제 기능 구현

na-young-kwon committed:
Merge branch 'step1' of https://github.com/jsim27/ios-cloud-notes into step1

jsim27 committed:
fix: 모든 셀을 지웠을 때 빈 메모를 표시하도록 수정

aCafela-coffee committed:
feat: DetailView에서 키보드가 화면을 가리지 않도록 수정

na-young-kwon committed:
feat: 셀 삭제, 추가시 애니메이션 효과 구현

jsim27 committed:
fix: 마지막 셀 삭제시 크래시되는 문제 해결

- 상속하지 않을 뷰컨트롤러 클래스 final 추가
- 셀 삭제 후 최상위 스크롤되지 않도록 수정

aCafela-coffee committed:
fix: 키보드가 활성화되고 다른 메모를 선택하면 스크롤되는 문제 해결

다른 메모를 선택하면 키보드를 비활성화 하도록 수정

na-young-kwon committed:
refactor: addObserver 메서드로 분리

- 수정되지 않는 tableView 상수로 변경

jsim27 committed:
refactor: delegate 프로토콜 네이밍 및 메서드명 변경

SungPyo commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    addObservers()
+  }
+  
+  @objc private func keyboardWillShow(_ notification: Notification) {
+    guard
+      let userInfo = notification.userInfo,
+      let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else {
+        return
+      }
+    textView.contentInset.bottom = keyboardFrame.height
+    textView.verticalScrollIndicatorInsets.bottom = keyboardFrame.height
+  }
```
요 함수에 있는 내용들을 고차함수를 이용해서 바꿔볼 수 있으세요? (숙제 아닙니다..)

aCafela-coffee replied:
```swift
private func addObservers() {
  let bottomInset = view.safeAreaInsets.bottom
  let addSafeAreaInset: (Notification) -> Void = { [weak self] notification in
    guard
      let self = self,
      let userInfo = notification.userInfo,
      let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else {
      return
    }
    self.additionalSafeAreaInsets.bottom = keyboardFrame.height - bottomInset
  }

  let removeSafeAreaInset: (Notification) -> Void = { [weak self] _ in
    self?.additionalSafeAreaInsets.bottom = 0
  }

  keyboardShowNotification = NotificationCenter.default.addObserver(
    forName: UIResponder.keyboardWillShowNotification,
    object: nil,
    queue: nil,
    using: addSafeAreaInset
  )
  keyboardHideNotification = NotificationCenter.default.addObserver(
    forName: UIResponder.keyboardWillHideNotification,
    object: nil,
    queue: nil,
    using: removeSafeAreaInset
  )
}
```
고차함수를 사용하도록 수정했습니다.

SungPyo commented on CloudNotes/CloudNotes/Controller/MemoListViewController.swift:
```diff
@@ -0,0 +1,130 @@
+  private let tableView = UITableView()
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    view.addSubview(tableView)
+    tableView.translatesAutoresizingMaskIntoConstraints = false
+    NSLayoutConstraint.activate([
+      tableView.topAnchor.constraint(equalTo: view.topAnchor),
+      tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+      tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+      tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+    ])
+    tableView.register(UITableViewCell.self, forCellReuseIdentifier: reuseIdentifier)
+    tableView.dataSource = self
+    tableView.delegate = self
+    setNavigationBar()
+    loadJSON()
+    let firstRowIndexPath = IndexPath(row: 0, section: 0)
+    tableView.selectRow(at: firstRowIndexPath, animated: false, scrollPosition: .top)
+    loadDetail(at: firstRowIndexPath)
```
함수로 목적따라 조금씩 나눠졌으면 좋겠어요.

aCafela-coffee replied:
```swift
override func viewDidLoad() {
  super.viewDidLoad()
  addTableView()
  setNavigationBar()
  loadJSON()
  tableView.selectRow(at: firstRowIndexPath, animated: false, scrollPosition: .top)
  loadDetail(at: firstRowIndexPath)
}
```

분리 했습니다.

SungPyo commented on CloudNotes/CloudNotes/SceneDelegate.swift:
```diff
@@ -7,48 +7,60 @@
+  func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
+    guard let windowScene = (scene as? UIWindowScene) else { return }
+    
+    let rootViewController = MemoListViewController()
```
이름보고 깜짝 놀랐어요 진짜 rootView인줄?

aCafela-coffee replied:
`memoListViewController` 로 변경했습니다.

SungPyo commented:
`커피, 나무, 조이 안녕하세요~ 첫 리뷰군요 :) 코드 자체는 큰 불편함 없이 잘 짜여진것 같아요! 질문하신 내용을 제외하면 특별하게 손 댈 부분도 현재 상황에선 보이지 않네요! 코드블럭에 코멘트를 조금 달아놓았는데 확인하시면 될 것 같구요 질문에 대한 답변은 해당 코멘트 하단에 달아두었습니다. 퀄리티는 이정도만 유지하면 충분해 보입니다. Step2에서 봬요 :)`

> 안녕하세요 웨더! **@SungPyo**
> 
> 수정이 필요한 부분들 한번씩 짚어주시면 감사하겠습니다 🙂 리뷰 잘 부탁드립니다~~ 🙏
> 
> ## 고민한 점
> ### 세로 화면에서 사이드바 사용하기
> ```swift
> // line 24, SceneDelegate.swift
> 
> let splitViewController = UISplitViewController(style: .doubleColumn)
> splitViewController.preferredDisplayMode = .oneBesideSecondary
> splitViewController.preferredSplitBehavior = .tile
> ```
> 
> 세로 화면에서 .oneBesideSecondary로 보여주기 위해 .tile을 사용했습니다.
> 
> ### 셀 삭제, 추가 애니메이션
> <img alt="" width="50%" height="50%" src="https://user-images.githubusercontent.com/70484506/153402759-e5fea7ef-7d96-4f50-9292-abbef415e622.gif">
> 
> ```swift
> // 셀 추가
> memos.insert(newMemo, at: 0)
> tableView.insertRows(at: [firstCellIndexPath], with: .fade)
> ```
> 
> ```swift
> // 셀 삭제
> memos.remove(at: indexPath.row)
> tableView.deleteRows(at: [indexPath], with: .fade)
> ```
> 
> ### 키보드가 디테일 뷰의 컨텐츠를 가리지 않을 것
> 키보드가 올라왔을 때 디테일 뷰의 내용을 가리지 않도록 텍스트뷰의 `contentInset` 값을 키보드의 높이만큼 더해주었습니다.
> 
> ```swift
> @objc private func keyboardWillShow(_ notification: Notification) {
>   guard
>     let userInfo = notification.userInfo,
>     let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else {
>       return
>     }
>   textView.contentInset.bottom = keyboardFrame.height
>   textView.verticalScrollIndicatorInsets.bottom = keyboardFrame.height
> }
> 
> @objc private func keyboardWillHide(_ notification: Notification) {
>   textView.contentInset.bottom = 0
>   textView.verticalScrollIndicatorInsets.bottom = 0
> }
> ```
> 
> <img alt="" width="50%" height="50%" src="https://user-images.githubusercontent.com/70484506/153402461-5f80d9ed-58db-4081-87b2-2d7d00e948b7.gif">
> 
> ### Memo.subtitle
> TableViewCell 두번째 줄의 레이블은 NSAttributedString를 이용했습니다.
> 
> ### 앱이 구동될 때 / 메모 삭제시 / 메모 추가 시 첫번째 셀 선택
> 앱이 처음 구동 될 때, 메모를 삭제할 때, 메모를 추가할 때 첫번째 셀을 선택하도록 구현했습니다. 사용자 입장에서 앱이 좀 더 편리하다고 느끼도록 했습니다.
> 
> ### 메모 목록 VC와 메모 상세 VC 간 데이터 전달 방식
> 메모 목록을 table로 가지고있는 `MemoListViewController` 와, 메모의 내용을 표시하는 `DetailViewController` 간에 양방향으로 delegation 관계를 구현하여 데이터를 전달하도록 했습니다.
> 
> `MemoListViewController`는 테이블뷰 셀이 선택되면 `MemoDisplayable` 프로토콜을 채택한 `DetailViewController` 에게 Memo 인스턴스를 전달하여 화면에 표시할 데이터를 전해줍니다.
> 
> `DetailViewController`는 사용자가 메모의 내용을 수정하면 `MemoStorable`을 채택한 `MemoListViewController` 에게 변경된 Memo 인스턴스를 전달하여 수정한 내용이 반영되도록 합니다.
> 
> ## 질문
> ### 셀 삭제시, 셀 선택이 취소됨
> ```swift
> func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
>   let deleteAction = UIContextualAction(style: .destructive, title: "Delete") { _, _, completionHandler in
>     self.memos.remove(at: indexPath.row)
>     if self.memos.isEmpty {
>       self.addMemo()
>     } else {
>       // 셀 삭제
>       tableView.deleteRows(at: [indexPath], with: .fade)
>       
>       // 가장 위의 셀 select
>       let firstRowIndexPath = IndexPath(row: 0, section: 0)
>       tableView.selectRow(at: firstRowIndexPath, animated: false, scrollPosition: .none)
>       
>       self.loadDetail(at: firstRowIndexPath)
>     }
>     completionHandler(true)
>   }
>   deleteAction.image = UIImage(systemName: "trash")
>   let configuration = UISwipeActionsConfiguration(actions: [deleteAction])
>   return configuration
> }
> ```
> 
> 테이블뷰의 셀을 스와이프하여 삭제한 후, 첫 번째 row의 cell을 select하는 기능을 위와 같이 구현하였지만 select되지 않는 문제가 있습니다.
> 
> selectRow 메서드의 부가적인 기능인 scrollPosition은 동작하는데 cell selection이 작동하지 않는 원인이 무엇인지 아직 알아내지 못했습니다.

ㄴ` 첫번째 인덱스를 선택하는 코드가 있네요. 해당 코드를 지우면 동작하지만 선택된 셀을 리스트에서 삭제하면 디테일 화면이 변경되지 않는 이슈가 있군요! 리스트에서 제거하는 셀이 현재 보고있는 셀과 동일한지만 체크한다면 될 것 같은데 어떠세요?`

> 
> ## 미해결
> ### 컬렉션뷰 및 DiffableDataSource 사용 실패
> 처음에 리스트 모양의 컬렉션뷰와 DiffableDataSource를 사용해서 MemoListViewController를 만들었습니다. 이때 수정한 메모를 저장하고, CollectionViewListCell에 반영하는 방법을 찾지 못했습니다. [commit 786d1e1c4966df3ad5ffbc635964bcb5191fef47](https://github.com/yagom-academy/ios-cloud-notes/pull/80/commits/786d1e1c4966df3ad5ffbc635964bcb5191fef47) UITableView로 변경했습니다.

ㄴ` 안타깝게도 저도 DiffableDataSource에 대한 지식이 전무합니다 :( (실무에서 사용중이 아니거든요.. 버전문제로 쓰지도 못합니다)
 리뷰어중에서는 올라프가 DiffableDataSource에 대한 지식이 해박하니 한번 물어보는건 어떨까요?`

jsim27 committed:
refactor: 키보드 Notification Observer 추가에 클로저 사용

jsim27 committed:
refactor: removeObservers함수 분리

aCafela-coffee committed:
feat: MemoListViewController에서 키보드가 컨텐츠를 가리지 않도록 수정

jsim27 committed:
refactor: MemoListViewController의 addTableView 함수 분리

jsim27 committed:
refactor: 의도치 않은 함수 분리 제거

aCafela-coffee committed:
refactor: firstRowIndexPath 프로퍼티 분리

aCafela-coffee committed:
refactor: memoListViewController 프로퍼티 이름 변경

jsim27 committed:
refactor: Array extension을 Collection Extension으로 변경

aCafela-coffee committed:
refactor: 선택된 셀 프로퍼티 추가, 셀 삭제시 선택되는 셀 변경

셀 삭제전의 선택되 있는 셀을 선택하도록 변경

jsim27 committed:
fix: ios15에서 자동완성이 잘못 적용되는 문제 해결

aCafela-coffee committed:
fix: TableView 셀 선택 관련 문제 해결

- TableView의 셀 선택이 풀리는 문제 해결
- 마지막 셀 삭제시 크래시 해결
- addObservers(), removeObservers() 함수가 여러번 호출될 때 문제가 없도록 변경
- MemoListViewController가 UITableViewController를 상속하도록 변경

jsim27 committed:
refactor: 키보드옵저버 제거 시점 변경
